main/report.go

```
@@ -0,0 +1,74 @@
+// package main contains function that calls the service N = 1000 times for each of the strings of the length 2,4,6,8,10.
+// Its calculate the percent of balanced strings for each length and print the results to stdout.
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+
+	"github.com/GoRunMaxim/Balanced-Parentheses/parentheses"
+)
+
+const (
+	serverCalls   = 1000 // amount of time that we call service
+	maxRequestNum = 100  // max request for server
+	generateURL   = "http://localhost:8080/generate?n="
+	channelBuffer = 10
+)
+
+func main() {
+	for i := 2; i <= 10; i += 2 {
+		percent := report(i)
+		fmt.Printf("Percent of balanced strings (length = %d) = %.1f\n", i, float64(percent)*100/serverCalls)
+	}
+}
+
+func report(length int) int {
+	var balancedPercent int
+	jobs := make(chan int, channelBuffer)
+	result := make(chan bool, channelBuffer)
+	for i := 0; i < maxRequestNum; i++ {
+		go worker(jobs, result)
+	}
+	go func() {
+		for i := 0; i < serverCalls; i++ {
+			jobs <- length
+		}
+		close(jobs)
+	}()
+	for i := 0; i < serverCalls; i++ {
+		if <-result {
+			balancedPercent++
+		}
+	}
+
+	return balancedPercent
+}
+
+func worker(length <-chan int, result chan<- bool) {
+	for v := range length {
+		if bracketStr, err := bracketsRequest(v); err == nil { // if we dont have error
+			result <- parentheses.Balance(bracketStr)
+		} else {
+			log.Println(err)
+		}
+	}
+}
+
+func bracketsRequest(length int) (string, error) {
+	resp, err := http.Get(generateURL + fmt.Sprint(length))
+	if err != nil {
+		return "", fmt.Errorf("Request error:" + err.Error())
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("Handler returned wrong status code: %v ", resp.StatusCode)
+	}
```

Let's return the error message that comes from the server.